codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,38 +177,77 @@ private static int compressionLevel(int blockSize) {
     }
 
     @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return  allocateBuffer(ctx, msg, preferDirect, true);
+    }
+
+    ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect,
+                                     boolean allowEmptyReturn) {
+        int targetBufSize = 0;
+        int remaining = msg.readableBytes() + buffer.readableBytes();
```

Comment:
We will get a more accurate `targetBufSize` if we include all the bytes in the calculation upfront; this is because we would potentially oversize a little extra.

Imagine this contrived example:

- `buffer`: capacity = 100, writeIndex = 70
- msg.readableBytes()` = 5

As we would execute a separate iteration of the loop for both `buffer` and `msg`, we would add:
- `70` and `HEADER_LENGTH` (which is 21) = 91 
- `5` and `HEADER_LENGTH` = 26
for a total of 115, which would cause use to allocate a buffer even though it will not be populated in `allocate()` as per my last round of changes.

If we add the number of bytes of `buffer` and `msg`before the loop (like I am doing), we get:
- `70` + `5` + HEADER_LENGTH = 96
and we would not allocate a buffer.

It's a little more efficient this way.

Now, wrt to overflow of just adding together `msg.readableBytes() + buffer.readableBytes()`,  you are still screwed as we still can only return a buffer from this function that is `Integer.MAX_VALUE` in size. Further, given that `compressor.maxCompressedLength()` returns a value that is larger than the input number, the situation will not improve. Thus, if anything, we can add a check after `int remaining = msg.readableBytes() + buffer.readableBytes()` to see if it overflowed or is above some value that we know will overflow.